examples/pytorch/graphsage/train_cv_multi_gpu.py

```diff
@@ -0,0 +1,409 @@
+            raise exception.__class__(trace)
+    return decorated_function
+
+def prepare_mp(g):
+    """
+    Explicitly materialize the CSR, CSC and COO representation of the given graph
+    so that they could be shared via copy-on-write to sampler workers and GPU
+    trainers.
+
+    This is a workaround before full shared memory support on heterogeneous graphs.
+    """
+    g.in_degree(0)
+    g.out_degree(0)
+    g.find_edges([0])
+
+def compute_acc(pred, labels):
+    """
+    Compute the accuracy of prediction given the labels.
+    """
+    return (th.argmax(pred, dim=1) == labels).float().sum() / len(pred)
+
+def evaluate(model, g, labels, val_mask, batch_size, device):
+    """
+    Evaluate the model on the validation set specified by ``val_mask``.
+    g : The entire graph.
+    inputs : The features of all the nodes.
+    labels : The labels of all the nodes.
+    val_mask : A 0-1 mask indicating which nodes do we actually compute the accuracy for.
+    batch_size : Number of nodes to compute at the same time.
+    device : The GPU device to evaluate on.
+    """
+    model.eval()
+    with th.no_grad():
+        inputs = g.ndata['features']
+        pred, _ = model.inference(g, inputs, batch_size, device)
+    model.train()
+    return compute_acc(pred[val_mask], labels[val_mask])
+
+def load_subtensor(g, labels, blocks, hist_blocks, dev_id, aggregation_on_device=False):
+    """
+    Copys features and labels of a set of nodes onto GPU.
+    """
+    blocks[0].srcdata['features'] = g.ndata['features'][blocks[0].srcdata[dgl.NID]].to(dev_id)
+    blocks[-1].dstdata['label'] = labels[blocks[-1].dstdata[dgl.NID]].to(dev_id)
+    for i, (block, hist_block) in enumerate(zip(blocks, hist_blocks)):
+        hist_col = 'features' if i == 0 else 'hist_%d' % i
+        block.srcdata['hist'] = g.ndata[hist_col][block.srcdata[dgl.NID]].to(dev_id)
+
+        # Aggregate history
+        hist_block.srcdata['hist'] = g.ndata[hist_col][hist_block.srcdata[dgl.NID]]
+        if aggregation_on_device:
+            hist_block.srcdata['hist'] = hist_block.srcdata['hist'].to(dev_id)
+        hist_block.update_all(fn.copy_u('hist', 'm'), fn.mean('m', 'agg_hist'))
+        block.dstdata['agg_hist'] = hist_block.dstdata['agg_hist']
+        if not aggregation_on_device:
+            block.dstdata['agg_hist'] = block.dstdata['agg_hist'].to(dev_id)
+
+def init_history(g, model, dev_id):
+    with th.no_grad():
+        history = model.inference(g, g.ndata['features'], 1000, dev_id)[1]
+        for layer in range(args.num_layers + 1):
+            if layer > 0:
+                hist_col = 'hist_%d' % layer
+                g.ndata['hist_%d' % layer] = history[layer - 1]
+
+def update_history(g, blocks):
+    with th.no_grad():
+        for i, block in enumerate(blocks):
+            ids = block.dstdata[dgl.NID]
+            hist_col = 'hist_%d' % (i + 1)
+
+            h_new = block.dstdata['h_new'].cpu()
+            g.ndata[hist_col][ids] = h_new
```

Comment:
Will there be data race issue?

Reply:
Yes.  In fact this operation is not even atomic.  Doesn't matter as I observed in both Reddit and OGB though.

Reply:
The children don't share the same frame apparently.